### DISCUSSION TITLE
Nextcloud (Alpine) vs NextcloudPi (Debian)

### DISCUSSION THREAD

**Discussion by maniac79** [tteck/Proxmox#3160] [Help Needed]
Hi,
first: I really love your script - such a blessing.

I spun up about 10-12 different LXCs successfully with your scripts and was easily able to setup many of them without issues for remote access behind my swag reverse proxy. I have swag running in Docker on a different host - that may or may not be important to know later?!

I'm having a hard time getting Nextcloud to run with remote access, though.
I first tried the Debian NextcloudPi LXC. Tried for a whole day, couldn't get it to run behind swag, it drove me mad, literally. :D
I wandered off to do other stuff and, everything else I tried afterwards, worked fine.
But I realy wanted to get Nextcloud running, so I tried again, this time with Alpine, out of curiosity.
Installed, configured and rest assured, after only a couple of minutes, it worked and I could even access it remotely.
And after only a few more minutes, I was even able to get rid of all the warnings it produced after first startup.
Bliss. For a moment. Because this remark bothers me regarding the Alpine installation:
"You are also obliged to adhere to updates provided by the package maintainer."
That is something that I'm personally note happy about. :-/

So, I thought, if I made a copy of the config.php, I could redo the Nextcloud Debian installation and I should be able to get it up and running now. I was confident, because I learned a lot, setting up the Alpine installation and other LXCs successfully.
And so I tried Debian again.
And whatever I do, I still can't access it remotely.
When I access it locally, everything seems to work fine, but as soon as I try to access it remotely, it just doesn't work.

Looking into the log files, I see entries such as:

```
[Sun Jun 09 18:00:19.623228 2024] [proxy_http:error] [pid 38554:tid 124906060646080] [client 127.0.0.1:42110] AH01114:  HTTP: failed to make connection to backend: 127.0.0.1

[Sun Jun 09 18:00:29.891022 2024] [proxy_http:error] [pid 38679:tid 124906148726464] [client 127.0.0.1:52042] AH01114: HTTP: failed to make connection to backend: 127.0.0.1

[Sun Jun 09 18:47:37.134152 2024] [ssl:error] [pid 488:tid 137769267967872] AH02217: ssl_stapling_init_cert: can't retrieve issuer certificate! [subject: CN=localhost / issuer: CN=localhost / serial: 68120D89F32...2DBEE3278 / notbefore: Oct 10 13:26:44 2023 GMT / notafter: Oct  7 13:26:44 2033 GMT]

[Sun Jun 09 18:47:37.134167 2024] [ssl:error] [pid 488:tid 137769267967872] AH02604: Unable to configure certificate localhost:443:0 for stapling
```

I think the "problem" that I have is that the Alpine install is Nextcloud and the Debian install is NextcloudPi.
The many different versions and installation methods of Nextcloud confuse me a lot, but I think NextcloudPi comes with something extra that I don't really need because I have my reverse proxy setup already up and running and that same thing is also what's blocking me to get it running.

Any help to get the Debian install working similar to the Alpine install is highly appreciated.
I even thought of starting from scratch via a plain Debian LXC, but I'm not sure if I could pull that off.

**Comment by Mellowlynx** [comment]
NextcloudPi works locally (i.a. using the LXC IP)? 
Does the NextcloudPi panel also work (lxc ip:4443)? If so, make sure you add the proxy IP as a trusted proxy.
It might also help to add the (sub)domain you use as trusted domain, in there too.

These are the only two things I changed, everything else I manage in the reverse proxy (NPM in my case)
I have no experience with swag, so sadly can't help you on that front.

**Reply by maniac79** [reply]
Hi,
yes: NextcloudPi panel works via ip:4443.
Subdomain is added as trusted domain.

Trusted domain and trusted proxies get automatically updated with every start of Nextcloud. When I remove stuff that's in there, to test if that has any effect, it gets overwritten and automatically added again.

This is how the trusted_domains are configured:
```
  'trusted_domains' => 
  array (
    0 => 'localhost',
    7 => 'nextcloudpi',
    5 => 'nextcloudpi.local',
    8 => 'nextcloudpi.lan',
    11 => '46.142.136.58',
    1 => '192.168.178.184',
    14 => 'nextcloud.fritz.box',
    3 => '<_sub.domain.tld_>',
  ),
```

I think I only added the sub.domain.tld myself via NextcloudPi, everything else, if I remember correctly, was auto populated.
But maybe I added 1-2 more. I couldn't find any way to check or delete any of these auto populated entries via the guy.
Same for the trusted proxies, which are configured like so:
```
  'trusted_proxies' => 
  array (
    11 => '127.0.0.1',
    12 => '::1',
    0 => '192.168.178.160:81',
    14 => '46.142.136.58',
  ),
```

The "192.168.178.168:81" is the only one which I added myself (I think via NextcloudPi) and that was the only entry that I had for the previously working Alpine installation. Hence I believe this should be all that's really required. But I can't even test removing the other entries, because something is adding them automatically back in.

I have started to setup a Docker installation in parallel on my Docker LXC, which also gives me headaches.
I don't know, I managed to get so many things up and running by now that I thought nothing could stop me.
Then along comes Nextcloud. 🥇

**Reply by Mellowlynx** [reply]
It's a bit difficult to troubleshoot, but with the info you gave, I would change:

Remove your wan IP from both lists, should not be needed, traffic is coming from your proxy. (and don't share it online)
Remove the port after the proxy IP.

If it works local, it's something between the proxy and the app, not the app.
As I have no experience with swag, I think you better can ask this question on a swag forum.

On a side note, you need to change your access security policy!
Nobody should be able to access your portainer remotely or your homarr.
Make a separation of what needs to be open to the public and IP restricts the rest. 
Use a VPN LXC to access the local stuff remotely.

Extra tip: Currently you have one SSL for all your subdomains. Try to separate them, and request one for only the subdomain.
Also, try to use a DNS proxy to hide your IP.

**Reply by maniac79** [reply]
Hi, the WAN IP is one of the things that I could not manually get removed. Whenever I remove it and restart Nextcloud, it gets added automatically again.

You're right regarding Portainer and Homarr... I tried to secure Portainer with Authentik, that didn't work very well. I might just cut it off as I don't see a scenario where I would need to access it remotely. Homarr I'm looking to replace soon.

"Try to separate them, and request one for only the subdomain." -> How would I do that?
"try to use a DNS proxy to hide your IP." -> Same question. :)

**Reply by Mellowlynx** [reply]
That is strange, my wan IP is not on my list, but is not part of your problem.

 - Nginx Proxy Manager does this for you, no idea if swag had that option.
    I also set up “Access Lists” in there. So from home or my work I can access it, but outside that, you can't.
 - Look at Cloudflare DNS, here you have that option.

**Reply by maniac79** [reply]
Managed to get the Docker container running, so I'll abandon setting it up as a separate LXC for the moment.
But thanks a lot for your help!
